### PR TITLE
Handle svelte label

### PR DESCRIPTION
**Pull request by unlocomqx**
Solves #7

### CODE REVIEW

**Comment by tomblachut** [comment]
I've been working on #4 recently because those issues are IMO related. First I've met the dead end but then asked JB and now I'm thinking your approach for this is in fact the best one.

Thanks for the tests, much appreciated.

I didn't run this yet but from the top of my head, have you tried using `HighlightingLexer` or `Annotator` instead of `HighlightingVisitor`?

https://intellij-support.jetbrains.com/hc/en-us/community/posts/206124159-Difference-between-HighlightVisitor-and-Annotator-

I've found this topic that says `Annotator` is recommended for plugins, on the other hand lexers are always the fastest but IDK if this is suitable

**Comment by unlocomqx** [comment]
I didn't try `Annotator` or `HighlightingLexer` but I will give them a try

**Comment by unlocomqx** [comment]
The annotator is more efficient indeed https://github.com/tomblachut/svelte-intellij/pull/43/commits/1f3643ae8310ff27a1872da00dda2da354139dc2

**Comment by tomblachut** [comment]
Nice, looks cleaner!

File: src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteAnnotator.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.lang.annotation.Annotator
+import com.intellij.lang.javascript.highlighting.JSHighlighter
+import com.intellij.lang.javascript.psi.JSLabeledStatement
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiElement
+import dev.blachut.svelte.lang.SvelteFileViewProvider
+
+class SvelteAnnotator : Annotator {
```
Comment (tomblachut) [inline src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteAnnotator.kt]: This annotator is buggy, doesn't highlight the label but random things after it.

image
image
image

File: src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+class SvelteReferenceContributor : PsiReferenceContributor() {
+    override fun registerReferenceProviders(registrar: PsiReferenceRegistrar) {
+        val pattern = PlatformPatterns.psiElement(JSReferenceExpression::class.java).and(FilterPattern(object : ElementFilter {
+            override fun isAcceptable(element: Any, context: PsiElement?): Boolean {
+                if (element is JSReferenceExpressionImpl) {
+                    if (DeclarationFinder.findDeclaration(element)) {
+                        return false
+                    }
+                    return PsiTreeUtil.findFirstParent(element) { parent -> parent is JSLabeledStatement } == null
```
Comment (tomblachut) [inline src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt]: I think this line prevents chained reactive declarations working.

Screenshot 2019-08-07 at 14 59 28

File: src/main/java/dev/blachut/svelte/lang/SvelteJsInspectionSuppressor.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+class SvelteJsInspectionSuppressor : InspectionSuppressor {
+    override fun getSuppressActions(element: PsiElement?, inspectionId: String): Array<SuppressQuickFix> {
+        return SuppressQuickFix.EMPTY_ARRAY
+    }
+
+    override fun isSuppressedFor(element: PsiElement, inspectionId: String): Boolean {
+        if (inspectionId == "UnnecessaryLabelJS") {
+            return element.parent is JSLabeledStatement && element.text == "$"
+        }
+        if (inspectionId == "JSUnusedAssignment") {
```
[the remark this answers is not in the document]
Comment (unlocomqx) [reply src/main/java/dev/blachut/svelte/lang/SvelteJsInspectionSuppressor.kt]: it accounts for this case, where the variable is assigned an initial value after the label
image

[the remark this answers is not in the document]
Comment (tomblachut) [reply src/main/java/dev/blachut/svelte/lang/SvelteJsInspectionSuppressor.kt]: Oh yeah, right. Is it even legal in Svelte?

File: src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt (hunk 2 above)
Comment (tomblachut) [reply src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt]: You mean `HighlightingVisitor`?

Comment (unlocomqx) [reply src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt]: I meant like `JSElementVisitor`

Comment (unlocomqx) [reply src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt]: I'm talking about this method to be exact
https://github.com/tomblachut/svelte-intellij/blob/1f3643ae8310ff27a1872da00dda2da354139dc2/src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt#L52
Do you think it's efficient like this? Would a `JSElementVisitor` be better in this case?

Comment (tomblachut) [reply src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt]: In the end visitor performs traversal so IDK similarly to find* methods, we would need to measure it, anyway I think it's okay for now. I'm gonna check your code again and merge it if nothing more turns up. Then I'm gonna try to tackle $ 
references.

Comment (unlocomqx) [reply src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt]: I just figured out that `PsiTreeUtil.findChildrenOfType` uses a recursive element visitor under the hood. So it's the same thing after all.

**Comment by tomblachut** [comment]
Annotator stil does not work for me. When I change info annotation to e.g. error colors kick in but with info severity it stays white.

Notice that I've extended TextRange to include colon and it is being highlighted.

Screenshot 2019-08-09 at 10 01 23

File: src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        }))
+        registrar.registerReferenceProvider(pattern, object : PsiReferenceProvider() {
+            override fun getReferencesByElement(element: PsiElement, context: ProcessingContext): Array<PsiReference> {
+                return arrayOf(SvelteLabeledReference(element, element.textRange))
+            }
+        })
+    }
+}
+
+class SvelteLabeledReference(element: PsiElement, textRange: TextRange) : PsiReferenceBase<PsiElement>(element, textRange), PsiPolyVariantReference {
```
Comment (tomblachut) [inline src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt]: I've removed `PsiPolyVariantReference` and code still works as far as I've checked. Am I missing something?

```kotlin
class SvelteLabeledReference(element: PsiElement, textRange: TextRange) : PsiReferenceBase<PsiElement>(element, textRange) {
    override fun resolve(): PsiElement? {
        val embeddedContent = PsiTreeUtil.findFirstParent(element) { psiElement -> psiElement is JSEmbeddedContent }
        val elementDefinition = PsiTreeUtil.findFirstParent(element) { psiElement -> psiElement is JSDefinitionExpression }
        val labeledStatements = PsiTreeUtil.findChildrenOfType(embeddedContent, JSLabeledStatement::class.java)
        labeledStatements.forEach {
            val definitions = PsiTreeUtil.findChildrenOfType(it, JSDefinitionExpression::class.java)
            definitions.filter { definition -> definition != elementDefinition }.forEach { definition ->
                if (definition.name == element.text && !DeclarationFinder.hasDeclaration(PsiTreeUtil.findChildOfType(definition, JSReferenceExpressionImpl::class.java))) {
                    return definition
                }
            }
        }
        return null
    }
}
```

Comment (unlocomqx) [reply src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteReferenceContributor.kt]: You're correct, it's a single reference so no need for `multiResolve` anyway, I will apply your code right away

**Comment by unlocomqx** [comment]
I was able to reproduce the annotator issue on the dracula theme
screenshot_2

Light theme gave expected results (I made keywords red for better visibility)
image

I will try to determine the cause and come up with a fix/workaround

**Comment by unlocomqx** [comment]
In the light theme, the label doesn't have a foreground color
image


In dracula, it has
image

Seems to be an issue of priority because the two annotations have severity = info
I removed the label annotation in https://github.com/tomblachut/svelte-intellij/pull/43/commits/00cdccc79cf927d52994655380ee2dafcfb0be10

**Comment by tomblachut** [comment]
Excellent work. I've never checked Color Scheme settings myself, seems you can use it for debugging 🥇

Sadly this code is so hacky I'm afraid it can stop working between IntelliJ releases. There was already such issue between 2019.1 and 2019.2 in this project. Anyway let's keep it that way for now.

I've experimented a bit and learned a bunch about annotations along the way. Nothing else worked lol.

Built in label annotation is added by `JSAnnotator`, which  indirectly calls `JSSemanticHighlightingUtil.class:590`. 

I've had some success by adding annotations to the same element as JSAnnotator does but there is race condition. I thought adding `order="last"` to the entry in `plugin.xml` will fix that but it doesn't work as I thought it does.

I have last clue remaining: `JSAnnotator` is created in `JSAnalysisHandlersFactory#createAnnotatingVisitor`. This is an extension point, so we could modify the method that creates annotations for labels instead of using separate Annotator. This would I think require creating new JS dialect and I suppose there will be complications. Asking JetBrains is the best course of action now, I'll do it later.

BTW This label is weird imo. 'let' and 'const' keywords don't have those.
image

I've remembered about reactive statements now. You are searching for definitions also in them so the following will find a reference. Such code will throw in runtime because Svelte treats JS as `strict mode` but more importantly IDE won't find the reference if you remove the label so I think we should handle it the same way.

image

Lastly references are being found inside of all `JSLabeledStatements`, not only Svelte ones.

One additional issue is duplicate finding. I think it should be made in separate PR.

image

**Comment by unlocomqx** [comment]
image

This labeling is added by 
https://github.com/tomblachut/svelte-intellij/blob/00cdccc79cf927d52994655380ee2dafcfb0be10/src/main/java/dev/blachut/svelte/lang/codeInsight/SvelteAnnotator.kt#L19

I could remove it if you like
I hope I understood you correctly regarding this point
I will check other points as well

**Comment by tomblachut** [comment]
I know. Yes, remove it please

**Comment by unlocomqx** [comment]
I removed it but other than that, I agree, it's hacky and probably many unexpected edge cases will arise on the ref contributor side

**Pull request closed without merging**